### DISCUSSION TITLE
Feature Suggestion: Booting from a zvol

### DISCUSSION THREAD

**Discussion by mjkl-gh** [zbm-dev/zfsbootmenu#498] [Feature Requests]
I've been trying to look into this, but unable to to find any documentation on how to set it up.

I did find the following post saying it should be possible: https://bbs.archlinux.org/viewtopic.php?id=208626

My usecase is as follows:

I've got a development machine setup with zfs. On top of the host I have some linux VM's I require for testing stuff (Not all in house software is equally supported between distro's) These vm's are installed onto seperate zvol's

Sometimes I have to visit the field and use the vm's to deploy software, I would prefer to be able to boot the vm's directly on the metal to not be dependent on the host configuration. Is this something that could be possible with ZBM and a usecase that would be supported?

**Comment by ahesford** [comment]
The kind of workflow you propose wouldn't lend itself to official support for a few reasons:
1. The zvol would need to contain a filesystem. If this filesystem were not ZFS, we would have to completely change our boot-environment search routines. Furthermore, having kernels on non-ZFS embedded in a zvol would break (or require a complete rewrite) a lot of the extra niceties exposed by our menu UI, like snapshot manipulation. (If you're putting a zpool on a zvol, you may be able to trick ZBM into doing what you want with an early-setup hook that pre-imports at least the base pool, or maybe all the pools on zvols.)
2. Booting from a zvol would require support inside the boot environment itself. Because no distribution supports this out of the box, you'll be writing a bunch of custom init logic for your environments. This would impose a lot of opinion on the identification of boot environments that we wouldn't want to officially endorse.
3. We have some additional logic, like exporting read-write pools during the tear-down before final boot, that assumes pools are independent. Supporting this like ZFS pools on zvols would require an understanding of additional dependencies that would add complexity just to support very rare uses.

With that said, it would certainly be cool to see something like this. Going one step further; I've often said that it would be neat to build a ZBM hypervisor that could simultaneously boot multiple environments from a pool. If you want to hack up a solution that would work for you, I'd recommend joining our IRC channel at `#zfsbootmenu` on Libera.chat. We might be able to provide some guidance on developing extensions to suit your needs in a manner that will be easy for you to maintain and extend.

**Reply by mjkl-gh** [reply]
Thanks for the detailed answer! Those are some good arguments why it will never work (officially)

Hacking together a custom solution does sound like something I would be interested in. However, You pointed out the complexity and the amount of work required to get it working, making it score lower on my priority list of things I want to work on. 

Hopefully someday...

Still thanks for the detailed answer!
